### PR TITLE
feat: ACI-1213 Use multipart upload

### PR DESCRIPTION
**Pull request by zsolt-marta-bitrise**

### CODE REVIEW

File: cache/network/api.go [hunk 1]
```diff
@@ -143,14 +139,56 @@ func (c apiClient) uploadArchive(archivePath, uploadMethod, uploadURL string, he
 	c.logger.Debugf("Response dump: %s", string(dump))
 
 	if resp.StatusCode != http.StatusOK {
-		return unwrapError(resp)
+		return "", unwrapError(resp)
+	}
+
+	etag := resp.Header.Get("ETag")
```
Comment (ofalvai) [inline cache/network/api.go:145]: Is it guaranteed that this header exists on the response?

File: cache/network/api.go [hunk 2]
```diff
@@ -143,14 +139,56 @@ func (c apiClient) uploadArchive(archivePath, uploadMethod, uploadURL string, he
...
+			c.logger.Warnf("chunk size mismatch, expected %d, got %d", chunkSize, len(chunkData))
+		}
+		if i == chunkCount-1 && len(chunkData) != lastChunkSize {
+			c.logger.Warnf("last chunk size mismatch, expected %d, got %d", lastChunkSize, len(chunkData))
+		}
+
+		c.logger.Debugf("Uploading chunk %d to %s", i, uploadURLs[i].URL)
+		etag, err := c.uploadArchiveChunk(uploadURLs[i], chunkData, int64(len(chunkData)))
+		if err != nil {
+			return nil, fmt.Errorf("upload chunk: %s", err)
```
Comment (ofalvai) [inline cache/network/api.go:182]: Can we also print the chunk index part of this error?

File: cache/network/api.go (hunk 1 above)
Comment (MrBlaise) [reply cache/network/api.go:145]: On S3 this is a must, on Google this won't be there AFAIK.

Comment (MrBlaise) [reply cache/network/api.go:145]: This Get method returns an empty string if there wasn't a header so this should work

File: cache/save.go [hunk 3]
```diff
@@ -7,17 +7,20 @@ import (
...
 	"github.com/bmatcuk/doublestar/v4"
 	"github.com/docker/go-units"
+
+	"github.com/bitrise-io/go-steputils/v2/cache/compression"
+	"github.com/bitrise-io/go-steputils/v2/cache/keytemplate"
+	"github.com/bitrise-io/go-steputils/v2/cache/network"
+	"github.com/bitrise-io/go-steputils/v2/stepconf"
 )
 
+const MultipartFileSizeLimit = 200 * 1024 * 1024 // MultipartFileSizeLimit above 200MB we upload using multi-part
```
Comment (rescuerunaways) [inline cache/save.go:22]: https://github.com/bitrise-io/advanced-build-cache-service/pull/110/files#r1387447015

it looks like there is mismatch in cache service?

File: cache/network/api.go [hunk 4]
```diff
@@ -143,16 +155,80 @@ func (c apiClient) uploadArchive(archivePath, uploadMethod, uploadURL string, he
...
+
+		if i < chunkCount-1 && len(chunkData) != chunkSize {
+			c.logger.Warnf("chunk size mismatch, expected %d, got %d", chunkSize, len(chunkData))
+		}
+		if i == chunkCount-1 && len(chunkData) != lastChunkSize {
+			c.logger.Warnf("last chunk size mismatch, expected %d, got %d", lastChunkSize, len(chunkData))
+		}
+
+		c.logger.Debugf("Uploading chunk %d to %s", i, uploadURLs[i].URL)
+		etag, err := c.uploadArchiveChunk(uploadURLs[i], chunkData, int64(len(chunkData)))
```
Comment (rescuerunaways) [inline cache/network/api.go:210]: do we plan to do it in parallel in future iterations?

File: cache/save.go (hunk 3 above)
Comment (zsolt-marta-bitrise) [reply cache/save.go:22]: Yeah, thanks for spotting. I think this is a leftover and can be removed. Initially, we wanted to call the new API only for multiparts, but we realized we can't do that as the backend needs to decide if the upload goes to GCS (which has no multipart support), so we moved all calls to the new API

File: cache/network/api.go (hunk 4 above)
Comment (zsolt-marta-bitrise) [reply cache/network/api.go:210]: Yeah it would be a great improvement later

**Comment by zsolt-marta-bitrise** [comment]
@rescuerunaways I see you have created a separate PR for the multi-threaded version. Do you intend to work on that so to close this?

**Pull request closed without merging**